dracut/repartition/endless-repartition.sh

```diff
@@ -95,26 +104,61 @@ if [ $added_space -gt 209715200 ]; then
   fi
 fi
 
-if [ $new_size -gt $part_size ]; then
-  # udev might still be busy probing the disk, meaning that it will be in use.
-  udevadm settle
+# udev might still be busy probing the disk, meaning that it will be in use.
+udevadm settle
+
+# take current partition table
+parts=$(sfdisk -d $root_disk)
+
+# check the last partition on the disk
+lastpart=$(echo "$parts" | sed -n -e '$ s/[^:]*\([0-9]\) :.*$/\1/p')
+
+if [ $lastpart -eq $swap_partno ]; then
+  # already have an extra partition, perhaps we were halfway through creating
+  # a swap partition but didn't finish. Remove it to try again below.
+  parts=$(echo "$parts" | sed '$d')
+elif [ $lastpart -gt $swap_partno ]; then
+  echo "repartition: found $lastpart partitions?"
+  exit 0
+fi
 
+# remove the last-lba line so that we fill the disk
+parts=$(echo "$parts" | sed -e '/^last-lba:/d')
+
+if [ $new_size -gt $part_size ]; then
   echo "Try to resize $root_part to fill $new_size sectors"
-  echo ",$new_size,," | sfdisk --force --no-reread -N2 $root_disk
-  echo "sfdisk returned $?"
-  udevadm settle
+  parts=$(echo "$parts" | sed -e "$ s/size=[0-9\t ]*,/size=$new_size,/")
 fi
 
 if [ -n "$swap_start" ]; then
   # Create swap partition
   echo "Create swap partition at $swap_start"
-  echo "$swap_start,+,S," | sfdisk --force --no-reread -N3 $root_disk
-  echo "sfdisk returned $?"
+  parts="$parts\nstart=$swap_start, type=0657FD6D-A4AB-43C4-84E5-0933C84B4F4F"
```

Comment:
Using `\n` in the assignment seems to work in dash but not in bash and I don't think it's specified behavior. I think it's safer to just continue on the next line:
```
parts="$parts
start=$swap_start, ..."
```
That's supported in any shell.